internal/controller/biossettings_controller.go

```diff
@@ -925,6 +925,42 @@ func (r *BiosSettingsReconciler) applyBiosSettingOnServer(
 	return err
 }
 
+func (r *BiosSettingsReconciler) ensureNoStrandedStatus(
+	ctx context.Context,
+	biosSettings *metalv1alpha1.BIOSSettings,
+) (bool, error) {
+	removeFlow := func(s []metalv1alpha1.BIOSSettingsFlowStatus, toRemove metalv1alpha1.BIOSSettingsFlowStatus) []metalv1alpha1.BIOSSettingsFlowStatus {
+		for index, item := range s {
+			if item.Name == toRemove.Name && item.Priority == toRemove.Priority {
+				return append(s[:index], s[index+1:]...)
+			}
+		}
+		return s
+	}
+
+	// Incase the settings Spec got changed during Inprogress and left behind Stale states clean it up.
+	settingsNamePriorityMap := map[string]int32{}
+	biosSettingsBase := biosSettings.DeepCopy()
+	biosFlowStateBase := append([]metalv1alpha1.BIOSSettingsFlowStatus{}, biosSettings.Status.FlowState...)
+	for _, settings := range biosSettings.Spec.SettingsFlow {
+		settingsNamePriorityMap[settings.Name] = settings.Priority
+	}
+	for _, flowStatus := range biosSettings.Status.FlowState {
+		if value, ok := settingsNamePriorityMap[flowStatus.Name]; !ok || value != flowStatus.Priority {
+			biosFlowStateBase = removeFlow(biosFlowStateBase, flowStatus)
+		}
+	}
```

Comment:
Now we got (roughly) quadratic complexity. :smile: What about filtering the invalid statuses?
```go
settingsNamePriorityMap := map[string]int32{}
for _, settings := range biosSettings.Spec.SettingsFlow {
		settingsNamePriorityMap[settings.Name] = settings.Priority
}
nextFlowStatuses := make([]metalv1alpha1.BIOSSettingsFlowStatus, 0)
for _, flowStatus := range biosSettings.Status.FlowState {
	if value, ok := settingsNamePriorityMap[flowStatus.Name]; ok && value == flowStatus.Priority {
		nextFlowStatuses = append(nextFlowStatuses, flowStatus)
	}
}
```